### PR TITLE
[2201.8.x] Fix performance drop drastically while profiling a Ballerina program

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/41300
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/41547

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by HindujaB** [comment]
Shall we add the 8.x branch to the pull_request workflows?
It seems they are not running.

**Review by gabilang — APPROVED** [review]
LGTM

File: bvm/ballerina-profiler/spotbugs-exclude.xml [hunk 1]
```diff
@@ -20,4 +20,8 @@
         <Class name="io.ballerina.runtime.profiler.Profiler"/>
         <Bug pattern="DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED"/>
     </Match>
+    <Match>
+        <Class name="io.ballerina.runtime.profiler.ui.JsonParser$Data"/>
+        <Bug pattern="URF_UNREAD_FIELD"/>
```
Comment (Nadeeshan96) [inline bvm/ballerina-profiler/spotbugs-exclude.xml:25]: Can't we remove the unread field from the class?

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/Data.java [hunk 2]
```diff
@@ -26,48 +27,33 @@
...
-    private long totalTime;
-    private long minTime;
-    private long maxTime;
-
-    public Data(String name) {
-        this.name = name;
+    public final String stackKey;
+    public final String stackIndex;
+    public String stackTrace = null;
+    public long totalTime;
```
Comment (Nadeeshan96) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/Data.java:33]: We can make these `protected`.

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/StackTraceMap.java [hunk 3]
```diff
@@ -29,24 +30,40 @@
...
-             if (stackTraceIndexMap.containsKey(stack)) {
-                 keyBuilder.append(stackTraceIndexMap.get(stack));
-             } else {
-                 stackTraceIndexMap.put(stack, localVarIndex);
-                 keyBuilder.append(localVarIndex++);
-             }
-         }
-        return keyBuilder.toString();
+    private static final ConcurrentHashMap<String, Integer> stackTraceIndexMap = new ConcurrentHashMap<>();
+    private static final ConcurrentHashMap<String, String> indexStackTraceMap = new ConcurrentHashMap<>();
```
Comment (Nadeeshan96) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/runtime/StackTraceMap.java:39]: Shall we move these before the constructor

File: bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/ui/FileUtils.java
Comment (Nadeeshan96) [inline bvm/ballerina-profiler/src/main/java/io/ballerina/runtime/profiler/ui/FileUtils.java:1]: Shall we add a private constructor for this class

**Review by Nadeeshan96 — COMMENTED** [review]
If possible shall we also add a test to check the performance of `bal profile` (maybe in another PR). We can use this same snippet https://github.com/ballerina-platform/ballerina-lang/issues/41300#issue-1869101439 (1st snippet here) and profile it and check whether it runs under 3 seconds or something like that. We can examine the output and check whether it contains all the user defined functions and reasonable times in the output.

When I tested the output I saw it does not contain `isSortedArray` function in the flame graph. How did that get omitted?
image

File: bvm/ballerina-profiler/spotbugs-exclude.xml (hunk 1 above)
Comment (warunalakshitha) [reply bvm/ballerina-profiler/spotbugs-exclude.xml:25]: This is intenral used in gson. Not used in profile code.

**Comment by warunalakshitha** [comment]
> If possible shall we also add a test to check the performance of `bal profile` (maybe in another PR). We can use this same snippet [#41300 (comment)](https://github.com/ballerina-platform/ballerina-lang/issues/41300#issue-1869101439) (1st snippet here) and profile it and check whether it runs under 3 seconds or something like that. We can examine the output and check whether it contains all the user defined functions and reasonable times in the output.
> 
> When I tested the output I saw it does not contain `isSortedArray` function in the flame graph. How did that get omitted? image

Will add a test in master PR.

**Review by Nadeeshan96 — APPROVED** [review]
LGTM

**Pull request merged**